kong/kong.lua

```diff
@@ -157,6 +161,34 @@ function Kong.ssl_certificate()
   end
 end
 
+function Kong.balancer()
+  local addr = ngx.ctx.balancer_address
+  addr.tries = addr.tries + 1
+  if addr.tries > 1 then  
+    -- only call balancer on retry, first one is done in `core.access.before` which runs
+    -- in the ACCESS context and hence has less limitations than this BALANCER context
+    -- where the retries are executed
+    
+    -- record failure data
+    addr.failures = addr.failures or {}
+    local state, code = ngx_balancer.get_last_failure()
+    addr.failures[addr.tries-1] = { name = state, code = code }
```

Comment:
Ok, was mostly wondering if this was implemented already somehow and if I missed it.

> Regarding logging; nginx will do the actual retry check and block if it hits it.

But since [proxy_next_upstream_retries](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_next_upstream_tries) is `0` (aka no limit) by default, when exactly will "nginx block"?

> So no extra logging required?

Why are we keeping track of the failures then?  Maybe don't bother? (actually need to be sure nginx will log those errors first).